sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVRelation.scala

```diff
@@ -195,18 +202,50 @@ private[sql] class CsvOutputWriter(
   private var records: Long = 0L
   private val csvWriter = new LineCsvWriter(params, dataSchema.fieldNames.toSeq)
 
-  private def rowToString(row: Seq[Any]): Seq[String] = row.map { field =>
-    if (field != null) {
-      field.toString
-    } else {
-      params.nullValue
+  private def rowToString(row: InternalRow): Seq[String] = {
+    var i = 0
+    val values = new Array[String](row.numFields)
+    while (i < row.numFields) {
+      if (!row.isNullAt(i)) {
+        values(i) = fieldsConverters(i).apply(row, i)
+      } else {
+        values(i) = params.nullValue
+      }
+      i += 1
+    }
+    values
+  }
+
+  private def makeConverter(dataType: DataType): ValueConverter = {
+    dataType match {
+      case DateType if params.dateFormat != null =>
+        (row: InternalRow, ordinal: Int) =>
+          params.dateFormat.format(DateTimeUtils.toJavaDate(row.getInt(ordinal)))
+
+      case DateType =>
+        (row: InternalRow, ordinal: Int) =>
+          DateTimeUtils.toJavaDate(row.getInt(ordinal)).toString
+
+      case TimestampType if params.dateFormat != null =>
+        (row: InternalRow, ordinal: Int) =>
+          params.dateFormat.format(DateTimeUtils.toJavaTimestamp(row.getLong(ordinal)))
+
+      case TimestampType =>
+        (row: InternalRow, ordinal: Int) =>
+          DateTimeUtils.toJavaTimestamp(row.getLong(ordinal)).toString
+
+      case udt: UserDefinedType[_] => makeConverter(udt.sqlType)
+
+      case dt: DataType =>
+        (row: InternalRow, ordinal: Int) =>
+          row.get(ordinal, dt).toString
```

Comment:
`schema` is being validated [here](https://github.com/apache/spark/blob/c775bf09e0c3540f76de3f15d3fd35112a4912c1/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVFileFormat.scala#L87). So it does not have to handle exception cases here again.